doc/api/cli.md

```diff
@@ -595,6 +595,31 @@ Path to the file used to store the persistent REPL history. The default path is
 `~/.node_repl_history`, which is overridden by this variable. Setting the value
 to an empty string (`''` or `' '`) disables persistent REPL history.
 
+### `NODE_UNHANDLED_REJECTION=state`
+<!-- YAML
+added: REPLACEME
+-->
+
+Setting this environment variable allows fine grained control over the behavior
+of unhandled rejections. This may be set to either one of `SILENT`, `WARN`,
+`ERROR_ON_GC` or `ERROR` while the default behavior without using the
+environment variable is the same as `WARN`. `WARN` reflects the
+[`unhandledRejection hook`][].
+
+Setting the environment variable to `SILENT` prevents unhandled rejections
+from being logged. The [`unhandledRejection hook`][] is still triggered in case
+it is set.
+
+If set to `ERROR_ON_GC` unhandled rejections that are garbage collected are
```

Comment:
What are the guarantees of this mode? It seems that there is no guarantee that the garbage collector will collect any rejected promise without a rejection handler even if there is no reference to them. Is that correct?

Reply:
Even if there was a guarantee that all rejected promises would be garbage collected at some point in the future, it seems it would make it difficult to reason about the behavior of a given process (e.g. determining whether a promise rejected that didn't have a catch handler) based on the exit code of that process.

For instance, the following code exits with an exit code of 0:

```
Promise.reject('Boom');

setTimeout(function logSomething() {
    console.log('logging something');
});
```
when run with `NODE_UNHANDLED_REJECTION=ERROR_ON_GC`.

However it could exit with a non-zero exit code too if the garbage collector kicks in before the timer expires and its underlying handle is closed.

Is that a problem?

Reply:
That is a problem we discussed and Ruben is going to follow up on this PR with a PR that checks if there are unhandled rejections pending when the process exits and then logs them and sets the status code.

So you're right and the behaviour should be to exit with a non-zero status code for it - but that's a different PR

Reply:
> That is a problem we discussed and Ruben is going to follow up on this PR with a PR that checks if there are unhandled rejections pending when the process exits and then logs them and sets the status code.

I don't think this would address all the concerns that I have. Let's use this code as an example:

```
Promise.reject(new Error('boom'));

setTimeout(function exitProcessAtSomePoint() {
  process.exit(0);
}, 2000);
```

What should happen in this case when running node with `UNHANDLED_REJECTION_ERROR=ERROR_ON_GC`? Would the exit code be guaranteed to be consistent depending on the GC's behavior?

> So you're right and the behaviour should be to exit with a non-zero status code for it - but that's a different PR

I'd rather see the details of the semantics of the `ERROR_ON_GC` mode be ironed out  before support for this mode lands. It seems that it could land separately from this PR anyway.

Reply:
@misterdjules 
> What should happen in this case when running node with UNHANDLED_REJECTION_ERROR=ERROR_ON_GC

It would likely end regularly with exit code 0. That is fine though as that mode is a best effort mode in which it is guaranteed that no false negative. I will outline that stronger in the documentation by providing examples for it.

> I'd rather see the details of the semantics of the ERROR_ON_GC mode be ironed out before support for this mode lands. It seems that it could land separately from this PR anyway.

I am not aware that there are any open details about the semantics of `ERROR_ON_GC` since this mode is limited in it's capability. I am strongly against leaving this mode out as this should become the default on newer Node.js versions. Otherwise it would never be possible to change the default to stricter version at all and IMO that would be a huge mistake.

Reply:
> It would likely end regularly with exit code 0. That is fine though as that mode is a best effort mode in which it is guaranteed that no false negative.

I'm still not sure if this wouldn't be the source of great confusion. Having the output of a program depend on when and how much the garbage collector collects seems like it could lead to programs having exit codes that change due to an external system that they don't have visibility/control on.

But maybe that's OK.

> I will outline that stronger in the documentation by providing examples for it.

Thank you, I think that'll help!

> I am strongly against leaving this mode out as this should become the default on newer Node.js versions

I think it'd be helpful to mention in the description of this PR that the goal is for "ERROR_ON_GC" to become the default mode in a future version of node.

> Otherwise it would never be possible to change the default to stricter version at all and IMO that would be a huge mistake.

Do you mean that the goal is to move the default mode to something like "ERROR" in the future? If so I think it'd also be helpful to mention that.